### PR TITLE
ImportDataSource - Add twitter typeahead to list servers

### PR DESCRIPTION
**Pull request by adube**
This PR introduces Twitter's typeahead in the ImportDataSource component.

The url is displayed in the suggestions, and used in the search.  This could be enhanced in the future (i.e. we could use the name instead / too).

## Todo

 * [x] review
 * [x] ~~validate CSS style - I "borrowed" the code from https://github.com/bassjobsen/typeahead.js-bootstrap-css/, is that okay?~~ Fixed by fred, thanks.
 * [x] find a solution about the fact that Typeahead only matches from the beginning of the searched string...
 * [x] requires master to build properly, see: #3071

### CODE REVIEW

File: contribs/gmf/src/directives/importdatasource.js [hunk 1]
```diff
@@ -167,6 +188,25 @@ gmf.ImportdatasourceController = class {
    */
   $onInit() {
     this.gmfExternalDataSourcesManager_.map = this.map;
+
+
+    if (this.serversEngine_) {
+      // Timeout to let Angular render the placeholder of the input properly,
+      // otherwise typeahead would copy the string with {{}} in it...
+      this.timeout_(() => {
+        const engine = goog.asserts.assert(this.serversEngine_);
```
Comment (fredj) [inline contribs/gmf/src/directives/importdatasource.js]: I find this a bit complicated and uneasy to read; what about simply have:
```
goog.asserts.assert(this.serversEngine_);
```
and use `this.serversEngine_.ttAdapter()` bellow

File: contribs/gmf/src/directives/importdatasource.js [hunk 2]
```diff
@@ -167,6 +188,25 @@ gmf.ImportdatasourceController = class {
...
+        const $urlInput = this.element_.find(
```
Comment (fredj) [inline contribs/gmf/src/directives/importdatasource.js]: can probably be simplified to `this.element_.find('input[name=url]');`

**Comment by fredj** [comment]
It doesn't seems to work for me; I've searched for `geo` or `swiss` in the input and had not response from typeahead

**Comment by adube** [comment]
>It doesn't seems to work for me; I've searched for geo or swiss in the input and had not response from typeahead

Typeahead only matches from the beginning of the string and doesn't seem to be configurable otherwise.

"That's by design. Since typeahead.js isn't a search engine, it should be okay if it uses a somewhat naive algorithm to get suggestions."

Any suggestion?

**Comment by adube** [comment]
Found a solution, trying it now...

https://stackoverflow.com/questions/22059933/twitter-typeahead-js-how-to-return-all-matched-elements-within-a-string

**Comment by adube** [comment]
@fredj Alright, I applied the solution it it worked.  Please, review...

**Comment by fredj** [comment]
Looks good to me. To have the same look between the search and the import data source (typeahead style), I've grouped the common css into a file. New commit to come

**Comment by fredj** [comment]
@adube see a91821cd61de3acf6d58704b3ad5d1c8e0bc9ba2

Does it look good to you?

**Comment by adube** [comment]
Code looks clean, however it changed the UI of the input in the ImportDS tool.  I've cherry-picked your commit and will make a quick fix right away.  Thank you very much for this.

**Comment by adube** [comment]
Okay, this looks perfect, now.

**Comment by adube** [comment]
@fredj Would you please take one last look at the last commit I made?  Just to make sure you agree with it.  If so, then I'll merge.

File: contribs/gmf/src/directives/importdatasource.js [hunk 3]
```diff
@@ -176,10 +176,11 @@ gmf.ImportdatasourceController = class {
          * Borrowed from:
          * https://stackoverflow.com/questions/22059933/twitter-typeahead-js-how-to-return-all-matched-elements-within-a-string
          *
-         * @param {string} url Url.
-         * @return {!Array.<string>} List of datum tokenizers.
+         * @param {BloodhoundDatum} url Url.
```
Comment (fredj) [inline contribs/gmf/src/directives/importdatasource.js]: the param name could be renamed to `datum` instead of `url`

**Comment by adube** [comment]
Build fails.  See: #3071

**Comment by fredj** [comment]
rebased

**Comment by adube** [comment]
Travis passes.  Merging now...

**Pull request merged**